awscli/customizations/logs.py

```diff
@@ -0,0 +1,321 @@
+
+
+class BaseLogEventsGenerator(object):
+    def __init__(self, client, timestamp_utils):
+        self._client = client
+        self._timestamp_utils = timestamp_utils
+
+    def iter_log_events(self, group_name, start=None, filter_pattern=None):
+        filter_logs_events_kwargs = self._get_filter_logs_events_kwargs(
+            group_name, start, filter_pattern)
+        log_events = self._filter_log_events(filter_logs_events_kwargs)
+        for log_event in log_events:
+            self._convert_event_timestamps(log_event)
+            yield log_event
+
+    def _filter_log_events(self, filter_logs_events_kwargs):
+        raise NotImplementedError('_filter_log_events()')
+
+    def _get_filter_logs_events_kwargs(self, group_name, start,
+                                       filter_pattern):
+        kwargs = {
+            'logGroupName': group_name,
+            'interleaved': True
+        }
+        if start is not None:
+            kwargs['startTime'] = self._timestamp_utils.to_epoch_millis(start)
+        if filter_pattern is not None:
+            kwargs['filterPattern'] = filter_pattern
+        return kwargs
+
+    def _convert_event_timestamps(self, event):
+        event['ingestionTime'] = self._timestamp_utils.to_datetime(
+            event['ingestionTime'])
+        event['timestamp'] = self._timestamp_utils.to_datetime(
+            event['timestamp'])
+
+
+class NoFollowLogEventsGenerator(BaseLogEventsGenerator):
+    def _filter_log_events(self, filter_logs_events_kwargs):
+        paginator = self._client.get_paginator('filter_log_events')
+        for page in paginator.paginate(**filter_logs_events_kwargs):
+            for log_event in page['events']:
+                yield log_event
+
+
+class FollowLogEventsGenerator(BaseLogEventsGenerator):
+    _TIME_TO_SLEEP = 1
+
+    def __init__(self, client, timestamp_utils, sleep=None):
+        super(FollowLogEventsGenerator, self).__init__(client, timestamp_utils)
+        self._sleep = sleep
+        if sleep is None:
+            self._sleep = time.sleep
+
+    def _filter_log_events(self, filter_logs_events_kwargs):
+        try:
+            for event in self._do_filter_log_events(filter_logs_events_kwargs):
+                yield event
+        except KeyboardInterrupt:
+            # The only way to exit from the --follow is to Ctrl-C. So
+            # we should exit the iterator rather than having the
+            # KeyboardInterrupt propogate to the rest of the command.
+            return
+
+    def _do_filter_log_events(self, filter_logs_events_kwargs):
+        last_event_ids = []
+        while True:
+            response = self._client.filter_log_events(
+                **filter_logs_events_kwargs)
+            for event in response['events']:
+                # For the case where we've hit the last page, we will be
+                # reusing the most recent nextToken to continue polling. This
```

Comment:
Does this mean we can see the same nextToken twicein a row?  Wouldn't we get a `PaginationError` when we use the botocore paginators in the NoFollowLogEventsGenerator class because of duplicate next tokens?

Reply:
Yes. Right now I'm not using the paginators in botocore. I have to do the pagination manually because paginators will stop after no pagination token is sent back in the response. Also based on the [docs](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html#CWL-FilterLogEvents-response-nextToken), it says that next tokens are vaild for 24 hours. So it should be fine to reuse them.

Reply:
> Yes. Right now I'm not using the paginators in botocore. 

I'm confused.  My question is about `NoFollowLogEventsGenerator`, which appears to be using the paginators (https://github.com/aws/aws-cli/pull/3729/files/1940387c79792b8b367618a8a87355fa02de14bb#diff-b85f7571d740176a2efa0d5e2e936d76R280).  Why isn't the duplicate next token not an issue in that case?